testing/integration/upgrade_fleet_test.go

```diff
@@ -0,0 +1,179 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build integration
+
+package integration
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/elastic-agent-libs/kibana"
+
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/pkg/testing/tools"
+	"github.com/elastic/elastic-agent/pkg/version"
+	"github.com/elastic/elastic-agent/testing/upgradetest"
+)
+
+// TestFleetManagedUpgrade tests that the build under test can retrieve an action from
+// Fleet and perform the upgrade. It does not need to test all the combinations of
+// versions as the standalone tests already perform those tests and would be redundant.
+func TestFleetManagedUpgrade(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		Stack: &define.Stack{},
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
+	startFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+	err = startFixture.Prepare(ctx)
+	require.NoError(t, err)
+	startVersionInfo, err := startFixture.ExecVersion(ctx)
+	require.NoError(t, err)
+
+	// Upgrade to a different build but of the same version (always a snapshot).
+	// In the case there is not a different build then the test is skipped.
+	// Fleet doesn't allow a downgrade to occur, so we cannot go to a lower version.
+	sameVersion := define.Version()
+	if !strings.HasSuffix(sameVersion, "-SNAPSHOT") {
+		sameVersion += "-SNAPSHOT"
+	}
+	endFixture, err := atesting.NewFixture(
+		t,
+		sameVersion,
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(t, err)
+	err = endFixture.Prepare(ctx)
+	require.NoError(t, err)
+	endVersionInfo, err := endFixture.ExecVersion(ctx)
+	require.NoError(t, err)
+	if startVersionInfo.Binary.Commit == endVersionInfo.Binary.Commit {
+		t.Skipf("Build under test is the same as the build from the artifacts repository (commit: %s)", startVersionInfo.Binary.Commit)
+	}
```

Comment:
Heads up: this does not work for repackaged agents (different agent package versions will have the same commit for the agent binary but different components).
For the moment, it will work

Reply:
How would that ever be the case for a PR? A PR should always result in a new build, and as a developer testing from your laptop it should always be from a custom build as well.

I don't see a reality where that is not true.

Reply:
If this is only for PRs run on CI it will work (because the merge commit is always different from whatever we may download from Artifact API)
In the case of a dev testing, it can happen once https://github.com/elastic/elastic-agent/issues/2579 and https://github.com/elastic/elastic-agent/issues/2780 are implemented (somebody is packaging uncommitted stuff on top of main and use a different AGENT_PACKAGE_VERSION value)

I agree it's not an immediate concern and the local case can be solved by a simple commit on whatever branch

Reply:
What about checking both the versions and the commit in the same compare, would that fix the compare when AGENT_PACKAGE_VERSION is set?

Reply:
Yep that would work nicely :+1: